File: bootstrap/src/args.cpp
#include "compiler.hpp"

#include <unordered_map>
#include <iostream>
#include <filesystem>

#include "boost/program_options.hpp"

#include "chaic.hpp"
#include "report.hpp"

namespace po = boost::program_options;

namespace chai {
    // OptionKind represents a command-line option kind: it is used to map
    // option names to integral values so they can be switched over.
    enum class OptionKind {
        ROOTDIR,  // <root directory>
        HELP,     // --help
        VERSION,  // --version
        DEBUG,    // --debug
        OUTPATH,  // --outpath
        OUTMODE,  // --outmode
        LOGLEVEL, // --loglevel
    };

    // A table mapping option names to integral option kinds.
    std::unordered_map<std::string, OptionKind> optionNameToKind {
        { "rootdir", OptionKind::ROOTDIR },
        { "help", OptionKind::HELP },
        { "version", OptionKind::VERSION },
        { "debug", OptionKind::DEBUG },
        { "outpath", OptionKind::OUTPATH },
        { "outmode", OptionKind::OUTMODE },
        { "loglevel", OptionKind::LOGLEVEL },
    };

    // A table mapping output mode names to OutputMode values.
    std::unordered_map<std::string, OutputMode> modeNameToEnum {
        { "exe", OutputMode::EXE },
        { "obj", OutputMode::OBJ },
        { "asm", OutputMode::ASM },
        { "llvm", OutputMode::LLVM },
    };

    // A table mapping log level names to LogLevel values.
    std::unordered_map<std::string, LogLevel> logLevelNameToEnum {
        { "silent", LogLevel::SILENT },
        { "error", LogLevel::ERROR },
        { "warn", LogLevel::WARN },
        { "verbose", LogLevel::VERBOSE },
    };

    bool Compiler::initFromArgs(int argc, char* argv[]) {
        // Build the command-line parser.
        po::options_description visibleDesc("usage: chaic [options] <root directory>\n\nOptions");

        // TODO Should we move this code elsewhere so we don't have a giant wall
        // of text sitting in the middle of this function?
        visibleDesc.add_options()
            ("help,h", "Display usage information (ie. this text).")
            ("version,v", "Displays the current compiler version.")
            ("debug,d", "Whether the compiler should output debug information.")
            ("outpath,o", po::value<std::string>()->default_value("out")->value_name("<output path>"), 
                "Sets the path for compilation output.  "
                "If a single output is to be produced, this should be a file.  "
                "If multiple outputs are to be produced, this should be a directory.  "
                "Default to out[.<platform-extension>] if unspecified."
            )
            ("outmode,m", po::value<std::string>()->default_value("exe")->value_name("<mode>"), 
                "Sets the compilation mode: the type of output the compiler should produce.  "
                "Valid values are:\n"
                "  - \"exe\" for producing an executable (default)\n"
                "  - \"obj\" for producing object files for each package\n"
                "  - \"asm\" for producing assembly files for each package\n"
                "  - \"llvm\" for producing LLVM IR source files for each package\n"
            )
            ("loglevel", po::value<std::string>()->default_value("verbose")->value_name("<log level>"),
                "Sets the compiler's log level.  Valid values are:"
                "  - \"verbose\" for outputting all messages (default)\n"
                "  - \"warn\" for outputing errors and warnings\n"
                "  - \"error\" for outputting errors only\n"
                "  - \"silent\" for no output"
            )
        ;

        po::options_description hiddenDesc;
        hiddenDesc.add_options()
            ("rootdir", po::value<std::string>(), "the root directory")
        ;

        po::positional_options_description positionalDesc;
        positionalDesc.add("rootdir", 1);

        po::options_description combinedDesc;
        combinedDesc.add(visibleDesc).add(hiddenDesc);

        // Run the parser on the command-line arguments.
        po::variables_map parseResult;
        try {
            po::store(
                po::command_line_parser(argc, argv)
                .options(combinedDesc)
                .positional(positionalDesc)
                .run(), 
                parseResult
            );
            po::notify(parseResult);
        } catch (po::unknown_option& uoe) {
            std::cout << "[error] unknown option: " << uoe.get_option_name() << '\n';
            return false; 
        }

        // Apply each of the arguments to the compiler.
        bool hasRootDir = false, hasOutPath = false;
        for (auto& option : parseResult) {
            auto kind = optionNameToKind[option.first];

            switch (kind) {
            case OptionKind::ROOTDIR:
                m_rootDir = std::filesystem::absolute(option.second.as<std::string>()).string(); 
                hasRootDir = true;
                break;
            case OptionKind::HELP:
                std::cout << visibleDesc << '\n';
                return false;
            case OptionKind::VERSION:
                std::cout << CHAI_NAME << " " << CHAI_VERSION << '\n';
                return false;
            case OptionKind::DEBUG:
                m_debug = true;
                break;
            case OptionKind::OUTPATH:
                m_outputPath = std::filesystem::absolute(option.second.as<std::string>()).string();
                hasOutPath = true;
                break;
            case OptionKind::OUTMODE:
            {
                std::string& mode = option.second.as<std::string>();

                if (modeNameToEnum.find(mode) == modeNameToEnum.end()) {
                    std::cout << "[error] invalid output mode\n";
                    return false;
                } else {
                    m_mode = modeNameToEnum[mode];
                }

                break;
            }  
            case OptionKind::LOGLEVEL:
                // TODO
                break;
            }
        }

        // Assert that the root directory has been specified.
        if (!hasRootDir) {
            std::cout << "[error] specify a root directory\n";
            return false;
        }

        // Set the default output path if it is unspecified.
        if (!hasOutPath) {
            m_outputPath = "out";

            #ifdef _WIN32
                if (m_mode == OutputMode::EXE)
                    m_outputPath += ".exe";
            #endif
        }

        // If we reach here then initialization was successful and we return true.
        return true;
    }
}

File: bootstrap/cmd/args.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"chaic/depm"
	"chaic/report"
	"chaic/util"
)

const usage = `Usage: chaic [flags|options] <path to root directory or file>

Flags:
------
-h, --help      Displays usage information (ie. this text).    
-v, --version   Displays the current compiler version.
-d, --debug     Whether the compiler should output debug information. 

Options:
--------
-o,  --outpath    Sets the path for compilation output.  If a single output is
                  to be produced, this should be a file.  If multiple outputs
                  are to be produced, then this should be a directory.  Defaults
                  to out[.<platform extension>] if unspecified.
-ll, --loglevel   Sets the compiler's log-level.  Valid values are:
                    - "verbose" for outputting all messages (default)
                    - "warn" for outputting errors and warnings
                    - "error" for outputting errors only
                    - "silent" for no output
-m, --outmode     Sets compiler's output mode.  Valid values are:
                    - "exe" for producing an executable (default)
                    - "lib" for producing a static library
                    - "dyn" for producing a dynamic/shared library
                    - "obj" for producing object files for each package
                    - "asm" for producing assembly file for each package
                    - "llvm" for producing LLVM IR for each package
`

// Prints the usage message an exits the compiler with the given exit code.
func printUsage(exitCode int) {
	fmt.Print(usage, "\n")
	os.Exit(exitCode)
}

// argParser is a command-line argument parser.
type argParser struct {
	// The arguments being parsed.
	args []string

	// The argument parser's position within those arguments.
	ndx int
}

// Set containing all the argument names that correspond to options.
var options = map[string]struct{}{
	"o":         {},
	"m":         {},
	"ll":        {},
	"-outpath":  {},
	"-outmode":  {},
	"-loglevel": {},
}

// argumentError displays an argument error and exits the program.
func argumentError(message string, args ...interface{}) {
	fmt.Print("argument error: ", fmt.Sprintf(message, args...), "\n\n")
	printUsage(1)
}

// nextArg parses the next command-line argument if one exists.  The first value
// is the name of the argument.  If this argument is positional, this value is
// empty.  The second value is the value of argument. If this value is empty,
// the argument is a flag.  If an argument exists, at least one of the returned
// values will be non-empty.  The final value indicates whether or not there was
// an argument to parse.
func (ap *argParser) nextArg() (string, string, bool) {
	if ap.ndx < len(ap.args) {
		arg := ap.args[ap.ndx]
		ap.ndx++

		if strings.HasPrefix(arg, "-") { // flag or option
			name := arg[1:]

			if _, ok := options[name]; ok { // option
				// Make sure the option value exists.
				if ap.ndx < len(ap.args) && !strings.HasPrefix(ap.args[ap.ndx], "-") {
					value := ap.args[ap.ndx]
					ap.ndx++
					return name, value, true
				} else {
					argumentError("option %s requires an argument", strings.TrimLeft(name, "-"))
				}
			} else { // flag
				return name, "", true
			}

		} else { // positional
			return "", arg, true
		}
	}

	// No arguments to parse.
	return "", "", false
}

// useArg attempts to use a single command-line argument to initialize the
// compiler.  If the argument is invalid, the program will exit.
func useArg(c *Compiler, name, value string) {
	switch name {
	case "h", "-help":
		printUsage(0)
	case "v", "-version":
		fmt.Println(util.ChaiCompilerID)
		os.Exit(0)
	case "d", "-debug":
		c.debug = true
	case "ll", "-loglevel":
		{
			var logLevel int
			switch value {
			case "silent":
				logLevel = report.LogLevelSilent
			case "error":
				logLevel = report.LogLevelError
			case "warn":
				logLevel = report.LogLevelWarn
			case "verbose":
				logLevel = report.LogLevelVerbose
			default:
				argumentError("invalid log level")
			}

			report.InitReporter(logLevel)
		}
	case "o", "-outpath":
		{
			absPath, err := filepath.Abs(value)
			if err != nil {
				argumentError("invalid output path: %s", absPath)
			}

			c.outputPath = value
		}
	case "m", "-outmode":
		{
			var mode int
			switch value {
			case "exe":
				mode = OutModeExecutable
			case "asm":
				mode = OutModeASM
			case "llvm":
				mode = OutModeLLVM
			case "obj":
				mode = OutModeObj
			case "lib", "dyn":
				argumentError("output mode not implemented yet")
			default:
				argumentError("invalid output mode")
			}

			c.outputMode = mode
		}
	case "":
		if c.rootPath == "" {
			absPath, err := filepath.Abs(value)
			if err != nil {
				argumentError("invalid root path: %s", value)
			}

			c.rootPath = absPath
		} else {
			argumentError("root path specified multiple times")
		}
	default:
		argumentError("unknown flag: %s", name)
	}
}

// NewCompilerFromArgs creates a new compiler instance based on the given
// command line arguments if the arguments are valid and compilation should
// continue: ie. if the user requests the compiler version, then compilation
// should not continue.
func NewCompilerFromArgs() *Compiler {
	c := &Compiler{
		outputMode: OutModeExecutable,
		depGraph:   make(map[uint64]*depm.ChaiPackage),
	}

	ap := argParser{args: os.Args[1:], ndx: 0}

	// Parse all command line arguments.
	for {
		if name, value, ok := ap.nextArg(); ok {
			useArg(c, name, value)
		} else {
			break
		}
	}

	// Check to make sure a root path was specified.
	if c.rootPath == "" {
		argumentError("a root path must be specified")
	}

	// Set default values for any optional unspecified flags.
	report.InitReporter(report.LogLevelVerbose)

	if c.outputPath == "" {
		c.outputPath = filepath.Join(c.rootPath, "out")
	}

	// Validate and correct the output path as necessary.
	switch c.outputMode {
	case OutModeASM, OutModeLLVM, OutModeObj:
		{
			finfo, err := os.Stat(c.outputPath)
			if err == nil {
				if !finfo.IsDir() {
					argumentError("output path must be a directory in the current output mode")
				}
			} else if !os.IsNotExist(err) {
				argumentError("invalid output path: %s", err)
			}
		}
	default:
		{
			finfo, err := os.Stat(c.outputPath)
			if err == nil {
				if finfo.IsDir() {
					argumentError("output path must be a file in the current output mode")
				}
			} else if !os.IsNotExist(err) {
				argumentError("invalid output path: %s", err)
			}

			if runtime.GOOS == "windows" && !strings.HasSuffix(c.outputPath, ".exe") {
				c.outputPath += ".exe"
			}
		}
	}

	// Load the Chai path.
	if chaiPath := os.Getenv("CHAI_PATH"); chaiPath != "" {
		c.chaiPath = chaiPath
	} else {
		report.ReportFatal("missing CHAI_PATH")
	}

	return c
}
